### PR TITLE
Add docs on customizable service containers [ONPREM-1025]

### PR DESCRIPTION
**Pull request by christian-stephen**
# Description
A brief description of the changes.

Add docs for container runner on setting up and using [customizable service containers](https://circleci.atlassian.net/l/cp/cwBe11DM).

# Reasons
A link to a GitHub and/or JIRA issue (if applicable).
Otherwise, a brief sentence about why you made these changes.

Jira: https://circleci.atlassian.net/browse/ONPREM-1025

# Content Checklist
Please follow our style when contributing to CircleCI docs. Our style guide is here: [https://circleci.com/docs/style/style-guide-overview](https://circleci.com/docs/style/style-guide-overview).

Please take a moment to check through the following items when submitting your PR (this is just a guide so will not be relevant for all PRs) 😸:

- [ ] Break up walls of text by adding paragraph breaks.
- [ ] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [ ] Keep the title between 20 and 70 characters.
- [ ] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [ ] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [ ] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [ ] Include relevant backlinks to other CircleCI docs/pages.

### CODE REVIEW

File: jekyll/_cci2/container-runner.adoc
Comment (christian-stephen) [inline jekyll/_cci2/container-runner.adoc:1]: There are a number of lint errors for preexisting docs (pretty sure I caught all new ones before opening the PR). Happy to address those in this PR if necessary.

File: jekyll/_cci2/container-runner.adoc [hunk 1]
```diff
@@ -95,6 +95,184 @@ agent:
...
+            resources:
+              requests:
+                cpu: "0.5"
+                memory: "200Mi"
+  resourceClasses:
+    your-namespace/your-resource-class:
+      serviceContainers:
+        exact:
+          - "cimg/postgres:16":
+            spec:
```
Comment (Stuart133) [inline jekyll/_cci2/container-runner.adoc:124]: This gets mapped straight onto a container struct, so we don't need `spec:`

Comment (christian-stephen) [reply jekyll/_cci2/container-runner.adoc:124]: This is what happens when you get ChatGPT to generate an example for you.

File: jekyll/_cci2/container-runner.adoc [hunk 2]
```diff
@@ -95,6 +95,177 @@ agent:
...
+      - "cimg/redis:6":
+        resources:
+          requests:
+            cpu: "0.5"
+            memory: "200Mi"
+  resourceClasses:
+    your-namespace/your-resource-class:
+      serviceContainers:
+        exact:
+          - "cimg/postgres:16":
```
Comment (Stuart133) [inline jekyll/_cci2/container-runner.adoc:122]: These go into a go map, so we don't want the `-`

This applies to all the match types.

I forgot to update the confluence doc & helm example, will do both now

Comment (christian-stephen) [reply jekyll/_cci2/container-runner.adoc]: Already fixed :)

**Review by Stuart133 — APPROVED** [review]

File: jekyll/_cci2/container-runner.adoc
Comment (rosieyohannan) [reply jekyll/_cci2/container-runner.adoc:1]: Fixed

**Comment by christian-stephen** [comment]
@rosieyohannan Is this good to go? Thanks 🙏

File: jekyll/_cci2/container-runner.adoc [hunk 3]
```diff
@@ -95,6 +95,177 @@ agent:
...
+          cpu: "0.6"
+          memory: "300Mi"
+    - name: cimg/mongo:5
+      resources:
+        requests:
+          cpu: "0.4"
+          memory: "150Mi"
+----
+
+In the following sections, we'll discuss these customization options in greater detail.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:195]:
```suggestion
In the following sections, we will discuss these customization options in greater detail.
```

File: jekyll/_cci2/container-runner.adoc [hunk 4]
```diff
@@ -95,6 +95,177 @@ agent:
           - name: "other"
 ```
 
+[#customizable-service-containers]
+=== Customizable service containers
+
+By default, link:https://circleci.com/docs/using-docker/#using-multiple-docker-images[service (or secondary) containers] inherit the same link:https://kubernetes.io/docs/tasks/configure-pod-container/[container configuration] as defined by the primary container. However, this behavior can be overridden using customizable service containers. This allows fine-tuned control over a service's resource usage on a per-image basis.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:101]:
```suggestion
By default, xref:using-docker#using-multiple-docker-images[service (or secondary) containers] inherit the same link:https://kubernetes.io/docs/tasks/configure-pod-container/[container configuration] as defined by the primary container. However, this behavior can be overridden using customizable service containers. Using the available overrides allows fine-tuned control over a service's resource usage on a per-image basis.
```

File: jekyll/_cci2/container-runner.adoc [hunk 5]
```diff
@@ -95,6 +95,177 @@ agent:
...
+----
+
+In this configuration:
+
+- `cimg/redis:6` matches the `exact` rule at the global scope (within `agent.serviceContainers`) and is allocated 0.5 CPU units and 200Mi of memory.
+- `cimg/postgres:16` matches the `exact` rule at the resource class scope (`your-namespace/your-resource-class`) and is allocated 1 CPU unit and 500Mi of memory.
+- `cimg/mysql:8` matches the `pattern` rule at the resource class scope and is allocated 0.6 CPU units and 300Mi of memory.
+- `cimg/mongo:5` doesn't match any rule from the service container options, hence defaults to the `default` rule at the resource class scope and is allocated 0.4 CPU units and 150Mi of memory.
+
+The rendered Pod specification would then appear as:
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:167]:
```suggestion
The rendered Pod specification would then appear as follows:
```

File: jekyll/_cci2/container-runner.adoc [hunk 6]
```diff
@@ -95,6 +95,177 @@ agent:
...
+----
+
+In the following sections, we'll discuss these customization options in greater detail.
+
+==== Image match types
+
+Image match types govern how images are matched for container customization. The types include:
+
+. *Exact:*
+For exact matching, the image string must be an exact match. For example, `cimg/redis:6.2.6` only matches the `cimg/redis:6.2.6` image.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:202]:
```suggestion
. *Exact:* For exact matching, the image string must be an exact match. For example, `cimg/redis:6.2.6` only matches the `cimg/redis:6.2.6` image.
```

File: jekyll/_cci2/container-runner.adoc [hunk 7]
```diff
@@ -95,6 +95,177 @@ agent:
...
+
+. *Prefix:*
+For prefix matching, the image string matches all images with a common prefix. For example, `cimg/redis:` will match any `cimg/redis` image regardless of the tag.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:205]:
```suggestion
. *Prefix:* For prefix matching, the image string matches all images with a common prefix. For example, `cimg/redis:` will match any `cimg/redis` image regardless of the tag.
```

File: jekyll/_cci2/container-runner.adoc [hunk 8]
```diff
@@ -95,6 +95,177 @@ agent:
...
+
+. *Pattern:*
+For pattern matching, a Go-based regex pattern is used to match images. For example, `cimg/(redis|postgres):.*` matches any `redis` or `postgres` image from the `cimg` repository regardless of the tag. Refer to the link:https://golang.org/pkg/regexp/syntax/[Golang regex syntax] and link:https://regex101.com/[regex101.com] to test your regular expressions.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:208]:
```suggestion
. *Pattern:* For pattern matching, a Go-based regex pattern is used to match images. For example, `cimg/(redis|postgres):.*` matches any `redis` or `postgres` image from the `cimg` repository regardless of the tag. Refer to the link:https://golang.org/pkg/regexp/syntax/[Golang regex syntax] and link:https://regex101.com/[regex101.com] to test your regular expressions.
```

File: jekyll/_cci2/container-runner.adoc [hunk 9]
```diff
@@ -95,6 +95,177 @@ agent:
...
+
+. *Default:*
+The `Default` match type applies when an image did not match any of the other image match types. It sets a single specification for all such service containers.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:211]:
```suggestion
. *Default:* The `Default` match type applies when an image did not match any of the other image match types. It sets a single specification for all such service containers.
```

File: jekyll/_cci2/container-runner.adoc [hunk 10]
```diff
@@ -95,6 +95,177 @@ agent:
...
+
+In terms of precedence, selectors follow the hierarchy: `Exact` -> `Prefix` -> `Pattern` -> `Default`. If a given image name doesn't match any rule in the hierarchy, it defaults to the `Default` rule. Note, match types defined at the resource class scope take precedence over those at the same match type.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:213]:
```suggestion
===== Order of precedence

Selectors follow the hierarchy: `Exact` -> `Prefix` -> `Pattern` -> `Default`. If a given image name does not match any rule in the hierarchy, it defaults to the `Default` rule. 

NOTE: Match types defined at the resource class scope take precedence over those at the same match type.
```

File: jekyll/_cci2/container-runner.adoc [hunk 11]
```diff
@@ -95,6 +95,177 @@ agent:
...
+
+==== Selection scope
+
+Selection scopes determine the context in which the customization is applied. This comprises:
+
+. *Resource class:*
+This scope specifies a custom configuration for all containers running within a particular resource class. For example, setting specific resources under `your-namespace/your-resource-class` impacts only the containers running within this specific class. This scope takes precedence over the Global scope.
+
+[source,yaml]
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:222]:
```suggestion
. *Resource class:* This scope specifies a custom configuration for all containers running within a particular resource class. For example, setting specific resources under `your-namespace/your-resource-class` impacts only the containers running within this specific class. This scope takes precedence over the Global scope.
+
[source,yaml]
```

File: jekyll/_cci2/container-runner.adoc [hunk 12]
```diff
@@ -95,6 +95,177 @@ agent:
...
+          resources:
+            requests:
+              cpu: "1"
+              memory: "500Mi"
+----
+
+. *Global:*
+This scope applies a custom configuration globally to all containers across all resource classes. It is considered when no matching scope is found at the Resource class level.
+
+[source,yaml]
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:238]:
```suggestion
. *Global:* This scope applies a custom configuration globally to all containers across all resource classes. It is considered when no matching scope is found at the resource class level.
+
[source,yaml]
```

File: jekyll/_cci2/container-runner.adoc [hunk 13]
```diff
@@ -95,6 +95,177 @@ agent:
...
+  serviceContainers:
+    exact:
+      "cimg/redis:6":
+        resources:
+          requests:
+            cpu: "0.5"
+            memory: "200Mi"
+----
+
+When it comes to precedence, the `Resource class` scope overrides any `Global` scope selection for a given match type. If a match is available in both scopes, the `Resource class` scope prevails.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:250]:
```suggestion
===== Order of precedence

The `Resource class` scope overrides any `Global` scope selection for a given match type. If a match is available in both scopes, the `Resource class` scope prevails.
```

**Review by rosieyohannan — APPROVED** [review]

**Pull request merged**